A bunch of changes to join v0.1 on top of #3 (#4)
apollo-specs/join: join.spec.md

  style core fill:none,stroke:fuchsia,color:fuchsia;
```

This document defines a [core feature](https://specs.apollo.dev/core/v0.1) named `join` for describing [core schemas](https://specs.apollo.dev/core/v0.1) which **join** multiple **subgraph** schemas into a single **supergraph** schema.

This specification provides machinery to:
- define [subgraphs](#def-subgraph) with the {join__Graph} enum and the {@join__graph} directive
- assign fields to subgraphs with {@join__field}
- declare additional data required and provided by subgraph field resolvers with the `requires` and `provides` arguments to {@join__field}

> - assign keys and ownership to types with {@join__type} and {@join__owner}

Reviewer: This bullet feels like it needs more context (the why behind keys and ownership) but none of the previous bullets do that. However, most of the previous bullets feel a bit more self-explanatory?

I haven't made up my mind on this suggestion, mostly just a thought worth considering.

Author: I think this section needs to be pretty succinct, but I added a link in 1d7d243 which maybe will help?

Reviewer: Yeah that works for me

# How to read this document

This document uses [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt) guidance regarding normative terms: MUST / MUST NOT / REQUIRED / SHALL / SHALL NOT / SHOULD / SHOULD NOT / RECOMMENDED / MAY / OPTIONAL.

## What this document isn't

> This document specifies only the structure and semantics of supergraphs. It's expected that a supergraph will generally be the output of a compilation process which composes subgraph schemas. The mechanics of that process are not specified normatively here. Conforming implementations may choose any approach they like, so long as the result conforms to the requirements of this document.

Reviewer: ```suggestion
This document specifies only the structure and semantics of supergraphs. It's expected that a supergraph will generally be the output of a compilation process which composes subgraphs. The mechanics of that process are not specified normatively here. Conforming implementations may choose any approach they like, so long as the result conforms to the requirements of this document.
```
Not your change, but...this feels redundant, and I think we want this to collapse to subgraph in the vernacular.

Author: Makes sense to me, for parallelism with "supergraph" at the beginning of the sentence.

<a name=def-composer>**Composers** (or **compilers**)</a> are producers which compose subgraphs into a supergraph. This document places no particular requirements on the composition algorithm, except that it must produce a valid supergraph.

<a namme=def-router>**Routers**</a> are consumers which serve a composed schema as a GraphQL endpoint. *This definition is non-normative.*
  - Graph routers differ from standard GraphQL endpoints in that they are not expected to resolve fields or communicate with (non-GraphQL) backend services on their own. Instead, graph routers receive GraphQL requests and service them by performing additional GraphQL requests. This spec provides guidance for implementing routers, but does not require particular implementations of query separation or dispatch, nor does it attempt to normatively separate routers from other supergraph consumers.

>   - Routers will often omit schema elements from the API schema that is used to validate client operations and that they present to clients via introspection ({join__Graph}, for example, will typically be omitted)

Reviewer: API schema should probably be defined if we're going to use it here. FWIW, I find the previous iteration plenty expressive and a bit less of a mouthful. Is "exposed via introspection" an incomplete description that this is needed?

It's obviously not wrong, but maybe redundant and the sentence may be getting a bit long to read.

Author: I linked API schema to the core spec. 9673ed4

I think it's important to be clear that the schema we're talking about is used for validation. Plenty of people turn off introspection; we want to make sure they understand that this is still a meaningful concept in that case.  (eg, as soon as we implement inaccessible fields this will be quite relevant.)

I agree this could be less of a mouthful while still containing both places that the API schema is used, though...

Author: Is a51b5a2a step in the right direction?

<a name=def-endpoint>**Endpoints**</a> are running servers which can resolve GraphQL queries against a schema. In this version of the spec, endpoints must be URLs, typically http/https URLs.

<a name=def-subgraph>**Subgraphs**</a> are GraphQL schemas which are composed to form a supergraph. Subgraph names and metadata are declared within the special {join__Graph} enum.

This spec does not place any requirements on subgraph schemas. Generally, they may be of any shape. In particular, subgraph schemas do not need to be supergraphs themselves or to follow this spec in any way; neither is it an error for them to do so. Composers MAY place additional requirements on subgraph schemas to aid in composition; composers SHOULD document any such requirements.

# Data Model

> TODO maybe this should be called something more like "Overview" or "Graph Routing in a Supergraph" or something?

Author: 6d9244d

> TODO maybe this should be called something more like "Overview" or "Graph Routing in a Supergraph" or something?

Author: d8f4687

*This section is non-normative.* It describes the motivation behind the directives defined by this specification.

> A supergraph schema describes a GraphQL schema that can be served by a router. The router does not contain logic to resolve any of the schema's fields; instead, the supergraph schema contain directives starting with {@join__} which tell the router which subgraph endpoints can resolve each field, as well as other information needed in order to construct subgraph queries.  The directives described in this specification are designed for a particular query planning algorithm, and so there are some restrictions on how they can be combined that originate from the requirements of this algorithm. (We hope that future versions of this specification can relax some of these restrictions.)

Author: I merged your change.

I separated the second half out into its own paragraph and added more color. But maybe it should just go away. I do think it helps to express the difference between "important restrictions" and "restrictions based on the current algorithm". 57ae8cf

> Each supergraph schema contains a list of the subgraphs. It represents this as an enum called [{join__Graph}](#join__Graph) with an enum value for each subgraph. Each enum value is annotated with a [{@join__graph}](#@join__graph) directive telling the router what endpoint can be used to reach the subgraph and a name for the subgraph that can be used in representations of query plans and diagnostic messages.

Author: I merged your commit.

I guess my point is saying that on the one hand, the name will be seen by actual humans, but on the other hand, it's not required to have a specific form or relate to anything else in particular (eg, the graphs themselves don't need to "know" their names). Any thoughts on making this more clear?

Uniqueness is mentioned normatively in the `@join__graph` section. 061aa3d tweaks the wording a bit

> Each supergraph schema contains a list of the subgraphs. It represents this as an enum called [{join__Graph}](#join__Graph) with an enum value for each subgraph. Each enum value is annotated with a [{@join__graph}](#@join__graph) directive telling the router what endpoint can be used to reach the subgraph and a name for the subgraph that can be used in representations of query plans and diagnostic messages.

Reviewer: I think naming it as a "human-readable identifier" tells me more clearly the point you're trying to get across and how it might be used.

..."that can be useful as a human-readable identifier in various use cases such as query plan visualization and server logs" (diagnostic messages is fine too)

> Each supergraph schema contains a list of the subgraphs. It represents this as an enum called [{join__Graph}](#join__Graph) with an enum value for each subgraph. Each enum value is annotated with a [{@join__graph}](#@join__graph) directive telling the router what endpoint can be used to reach the subgraph and a name for the subgraph that can be used in representations of query plans and diagnostic messages.

Author: c3d8d12 on main

> To resolve a field, the router needs to know to which subgraphs it can delegate the field's resolution. The most direct way to indicate this in a supergraph schema is by annotating the field with a [{@join__field}](#@join__field) directive specifying which subgraph should be used to resolve that field.

Reviewer: >most direct way

Is there another way? The language might be a bit misleading

Author: "direct" might be the wrong word here. But there are other ways — it can be on a value type, it can be `provides`ed by the parent field, it can be implicit from the owner of the type.  Is 2d318dd better?

> In order for the router to send an operation that resolves a given field on a parent object to a subgraph, the operation needs to first resolve the parent object itself. There are several ways to accomplish this, described below. The examples below include abbreviated versions of the supergraph schemas which do not include the `schema` definition, directive definitions, or the `join__Graph` definition. This specification does not require the subgraph queries to be the same as those described in these examples; this is just intended to broadly describe the meanings of the directives.

Reviewer: Below we use the term "subgraph operation" which I like. Can we swap all usages of "subgraph query/queries" with that for consistency?

Author: 4ea27c2

## Root fields

> If this field appears at the root of the overall operation (query or mutation), then the field can just be placed at the root of the subgraph operation.

Reviewer: "this" field reads a little funny - which field?
```suggestion
If a field appears at the root of the overall operation (query or mutation), then it can be placed at the root of the subgraph operation.
```

Author: Applied. Now I'll go to the same to the other sections.

Author: aab84d4

> ```graphql example -- Root fields

Reviewer: Should we also demonstrate combining root fields from the same subgraph here?

i.e.
```
## On A:
{ fieldA fieldAlsoFromA }
```

Author: fb85cb4

# Supergraph schema
type Query {
  fieldA: String @join__field(graph: A)
  fieldB: String @join__field(graph: B)
}

# Operation
{ fieldA fieldB }
# Generated subgraph operations
## On A:
{ fieldA }
## On B:
{ fieldB }
```

## Fields on the same subgraph as the parent operation

> If this field's parent field will be resolved by an operation on the same subgraph, then this field can be resolved as part of the same operation, by putting it in a nested selection set on the parent field's subgraph operation. Note that this example contains a {@join__owner} directive on an object type; this will be described later.

Reviewer: "this" again feels strange to me
```suggestion
If a field's parent field will be resolved by an operation on the same subgraph, then it can be resolved as part of the same operation, by putting it in a nested selection set on the parent field's subgraph operation. Note that this example contains a {@join__owner} directive on an object type; this will be described later.
```

Author: heh, already applied in aab84d4

```graphql example -- Fields on the same subgraph as the parent operation
# Supergraph schema
type Query {
  fieldA: X @join__field(graph: A)
}

> type X @join__owner(graph: A) {

Reviewer: Are we intentionally providing an invalid example here? Seems plausible due to unintroduced concepts or for brevity.

`type X` should have at least one `@join__type`

Author: No, good catch. I think I convinced myself it wasn't required. 9058845

Reviewer: Good catch on the surrounding language updates, too 👍

  nestedFieldA: String @join__field(graph: A)
}

# Operation
{ fieldA { nestedFieldA } }
# Generated subgraph operations
## On A:
{ fieldA { nestedFieldA }}
```

## Fields provided by the parent field

> When resolving a field to a composite type (object, interface, or union), it is sometimes easy to calculate the values of nested fields inside that object even if those nested fields are ordinarily resolved in a different subgraph. When this is the case, you can include a `provides` argument in the `@join__field` listing these "pre-calculated" fields. The router can now resolve these fields in the "providing" subgraph instead of in the subgraph that would usually be used to resolve those fields.

Reviewer: ```suggestion
Sometimes, a subgraph may be capable of resolving nested fields (which it does not contribute to the graph) on a type. When this is the case, you can include a `provides` argument in the `@join__field` listing these "pre-calculated" fields. The router can now resolve these fields in the "providing" subgraph instead of in the subgraph that would usually be used to resolve those fields.
```

I honestly don't know how to reword this but this is where I'm at. I think the second half of what you have reads very well. The first is a bit confusing (and clearly difficult to articulate well, since I'm struggling to come up with something that I'm happy with).

Author: Hmm. I agree that my wording was confusing but I am not sure about your one either. Maybe an example would help? d010a05 tries a different approach for the first sentence plus a more motivating example.

Reviewer: This is a huge improvement! Really like how the explanation turned out (and the examples are great too)

```graphql example -- Provided fields
# Supergraph schema
type Query {
  fieldA: X @join__field(graph: A, provides: "usuallyBField")
  fieldB: X @join__field(graph: B)
}

> type X @join__owner(graph: B) {

Reviewer: Missing `@join__type` here too

Author: done previously

  usuallyBField: String @join__field(graph: B)
}

# Operation
{ fieldB { usuallyBField } }
# Generated subgraph operations
## On B
{ fieldB { usuallyBField } }

# Operation
{ fieldA { usuallyBField } }
# Generated subgraph operations
## On A

> { fieldA { usuallyBField } }

Reviewer: Differentiating these example operations with explanatory words could be helpful.

i.e.
typical resolution path
resolution path leveraging provides

Author: 474f882

```

## Fields on value types

> Some types have the property that all of their fields can be resolved by *any* subgraph that can resolve a field returning that type. These types are called *value types*. (Imagine a type `type T { x: Int, y: String }` where every resolver for a field of type `T` actually produces an object like `{x: 1, y: "z"}`, and the resolvers for the two fields on `T` just unpack the values already in the object.) In a supergraph schema, a type is a value type if it does not have a [{@join__owner}](#@join__owner) directive on it.

Reviewer: Do we want to collapse "supergraph schema" to just "supergraph" everywhere?

Author: I think all the places I'm saying "supergraph" are talking about the actual schema document (whereas "supergraph" can also mean the server, the abstract concept of the data graph, etc) so I think it's an OK distinction?

Reviewer: Aren't we talking about the document in this case (if we're talking about where directive usages are applied, etc.)

Author: Oops, typo in my last comment: I meant to say 

> all the places I'm saying "supergraph schema" are talking about the actual schema document

Does that make more sense?

```graphql example -- Value types
# Supergraph schema
type Query {
  fieldA: X @join__field(graph: A)
  fieldB: X @join__field(graph: B)
}

type X {
  anywhere: String
}

# Operation
{ fieldA { anywhere } }
# Generated subgraph operations
## On A
{ fieldA { anywhere } }

# Operation
{ fieldB { anywhere } }
# Generated subgraph operations
## On B
{ fieldB { anywhere } }
```

## Owned fields on owned types

> We've finally reached the most interesting case: a field that must be resolved by an operation on a different subgraph from the subgraph on which its parent field was resolved. In order to do this, we need a way to tell the subgraph to resolve that parent object. We do this by defining a special root field in the subgraph's schema: `Query._entities(representations: [_Any!]!): [_Entity]!`. This field takes a list of "representations", which are JSON objects containing a `__typename` field and some other fields, and returns a list of the same length of the corresponding objects resulting from looking up the representations in an application-dependent way.

Reviewer: This is a lot to drop on someone. What's an `_Any` and `_Entity`? Should we clarify that "some other fields" are the keys required to resolve the entity?

I'm also not sure if it belongs in the spec. I think there is a disclaimer about how the examples provided are not necessarily prescriptive, but this is an implementation detail of specifically how Apollo Federation works. This is getting into the details of query planning / execution which I'm not sure this spec is meant to be responsible for.

This is totally up for discussion but wanted to express my initial reaction. I understand that our implementation is useful in demonstrating / explaining the concepts. If we decide we want to do that, a note beforehand would substantially quiet my concerns.

Author: So I get the part that's "it doesn't have to be Apollo Federation" but I interpreted that to mean "you could have just written this supergraph schema by hand, there's no promise that there were subgraph schemas with `@key` on it".

I don't see how this spec describes something that could be used to drive a router other than by using `Query._entities`? I mean what else are you supposed to do with `key` and `requires`? Certainly I can imagine a future where the query isn't hardcoded and there are other ways to get non-root fields from subgraphs, though that would have to be an extension of the format?  Like, it does feel to me that `Query._entities` is a hard-coded part of the "router algorithm" for this version of the join spec (note that it was in the part that I deleted too).

I gave restructuring this a bit of a shot in d6c90c0. Not sure if it helps

```graphql example -- Value types
# Supergraph schema
type Query {
  fieldA: X @join__field(graph: A)
  fieldB: X @join__field(graph: B)
}

type X {
  anywhere: String
}

# Operation
{ fieldA { anywhere } }
# Generated subgraph operations
## On A
{ fieldA { anywhere } }

# Operation
{ fieldB { anywhere } }
# Generated subgraph operations
## On B
{ fieldB { anywhere } }
```

## Owned fields on owned types

We've finally reached the most interesting case: a field that must be resolved by an operation on a different subgraph from the subgraph on which its parent field was resolved. In order to do this, we need a way to tell the subgraph to resolve that parent object. We do this by defining a special root field in the subgraph's schema: `Query._entities(representations: [_Any!]!): [_Entity]!`. This field takes a list of "representations", which are JSON objects containing a `__typename` field and some other fields, and returns a list of the same length of the corresponding objects resulting from looking up the representations in an application-dependent way.

> There are several ways that the router can calculate a representation to pass to a subgraph. In this specification, all non-value types have a specific subgraph referred to as its "owner", specified via a `@join__owner` directive on the type. (Object types that are not value types are referred to as "entities"; the type `_Entity` referenced above is a union defined in each subgraph's schema consisting of the entity types defined by that subgraph. Only subgraphs which define entities need to define the `Query._entities` field.) The type must also have at least one `@join__type` directive specifying the owning subgraph along with a {key}. For each additional subgraph which can resolve fields returning that type, there should be exactly one `@join__type` directive specifying that subgraph along with a {key}, which should be identical to one of the keys specified with the owning subgraph.

Author: I adjusted the parens a couple commits ago

A key is a set of fields on the type (potentially including sub-selections and inline fragments), specified as a string. If a type `T` is annotated with `@join__type(subgraph: G, key: "a b { c }")`, then it must be possible to resolve the full field set provided as a key on subgraph G. Additionally, if you take an object with the structure returned by resolving that field set and add a field `__typename: "T"`, then you should be able to pass the resulting value as a representation to the `Query._entities` field on subgraph G.

> In order to resolve a field on an entity on the subgraph that owns its parent type, where that subgraph is different from the subgraph that resolved its parent object, the router first resolves a key for that object on the previous subgraph, and then uses that representation on the owning subgraph.

Reviewer: This sentence is really difficult to reason about. I'm trying to keep track of a field, an entity, two subgraphs, and a parent type and how they all relate to one another in my head. I think you're explaining the condition which results in needing an `entities` fetch, is that right?

I'm having a hard time putting this into words myself, can we illustrate by example instead?
"Consider the following arrangement..."
"Consider this operation, and specifically this nested field, ..."
"Notice the subgraph ownership changes here, which means we need to perform an entities query"

Or at least getting this into bite size things that will make it simpler for me to visualize crossing the boundary and the arrangement it takes to get to that condition.

Author: I'm going to come back to this one because I'm running low on time and it's hard

> For convenience, you may omit `@join__field(graph: A)` directives on fields whose parent type is owned by `A`.

Reviewer: This is a @martijnwalraven question, but I'm not sure if we permit this behavior or not. Right now composition does not omit these, so I don't know the answer to this.

Author: That's my interpretation based on this in buildQueryPlan.ts:

```ts
  getOwningService(
    parentType: GraphQLObjectType,
    fieldDef: GraphQLField<any, any>,
  ): string | undefined {
    return (
      getFederationMetadataForField(fieldDef)?.graphName ??
      this.getBaseService(parentType)
    );
  }
```

Let's leave this here for now but @martijnwalraven can correct us eventually if it's wrong?

```graphql example -- Owned fields on owned types
# Supergraph schema
type Query {
  fieldB: X @join__field(graph: B)
}

> type X

Reviewer: Are we interested in consistent (prettier) formatting on these?

Author: That would be nice but not sure I'm going to figure that out today (and can always be changed later)

    @join__owner(graph: A)
    # As the owner, A is allowed to have more than one key.
    @join__type(graph: A, key: "x")
    @join__type(graph: A, key: "y z")
    # As non-owners, B and C can only have one key each and
    # they must match a key from A.
    @join__type(graph: B, key: "x")
    @join__type(graph: C, key: "y z")
  {

>   # Because B owns X, we can omit @join__field(graph: B)
>   # from these two fields.

Reviewer: ```suggestion
  # Because A owns X, we can omit @join__field(graph: A)
  # from these three fields.
```

  x: String
  y: String
  z: String

> }

Reviewer: Out of fear of the example growing too large, it would be good to see fields owned by B and/or C as well here.

Reviewer: This is resolved in a more complex example below.

Author: Yep, this whole section is the "owned fields on owned types" section

In order to use the directives described by this specification, GraphQL requires you to include their definitions in your schema.

Processors MUST validate that you have defined the directives with the same arguments, locations, and `repeatable` flag as given below.

:::[definition](join.spec.graphql)

Processors MUST validate that the schema contains an enum named {join__Graph}; see [its section below](#join__Graph) for other required properties of this enum.

As described in the core specification, all of the directives and enums defined by this schema should be removed from the supergraph's [API schema](https://specs.apollo.dev/core/v0.1/#sec-Parts-of-a-Core-Schema). For example, the {join__Graph} enum should not be visible via introspection.

TODO define field set somewhere?

# Enums

##! join__Graph

Enumerate subgraphs.

```graphql definition
enum join__Graph
```

Documents MUST define a {join__Graph} enum. Each enum value describes a subgraph. Each enum value MUST have a [{@join__graph}](#@join__graph) directive applied to it.

> :::[example](photos.graphql#join__Graph) -- Using join__Graph to define endpoints

Reviewer: ```suggestion
:::[example](photos.graphql#join__Graph) -- Using join__Graph to define subgraphs and their endpoints
```

<a namme=def-router>**Routers**</a> are consumers which serve a composed schema as a GraphQL endpoint. *This definition is non-normative.*
  - Graph routers differ from standard GraphQL endpoints in that they are not expected to resolve fields or communicate with (non-GraphQL) backend services on their own. Instead, graph routers receive GraphQL requests and service them by performing additional GraphQL requests. This spec provides guidance for implementing routers, but does not require particular implementations of query separation or dispatch, nor does it attempt to normatively separate routers from other supergraph consumers.

>   - Routers expose an [API schema](https://specs.apollo.dev/core/v0.1/#sec-Parts-of-a-Core-Schema) to clients that is creating by transforming the supergraph schema (for example, the {join__Graph} enum and the directives described in this spec are removed from the API schema). The API schema is used to validate client operations and may be exposed to clients via introspection.

Reviewer: I like this a lot more.
>to clients that is creating by transforming
->
>to clients that is created by transforming

Author: 5de6c32

  fieldB: X @join__field(graph: B)
}

> type X @join__owner(graph: B) @join__type(graph: B, key: "usuallyBField") {

Reviewer: Something about providing a key doesn't seem right to me. It might be fine, it's also an interesting possible edge case with possibly undefined behavior (in our implementation) that I haven't thought about before.

I don't have an answer for this right now but I think it would be good for us to write an issue to investigate and test this. In the mean time, can we use an example that doesn't provide a field which is a key?

cc @martijnwalraven

Author: I actually already changed this to a different example and I'm providing a different field now too :)
